crates/turbo-tasks-memory/src/task_stats.rs

```diff
@@ -0,0 +1,114 @@
+use std::{
+    sync::atomic::{AtomicBool, Ordering},
+    time::Duration,
+};
+
+static ENABLE_FULL_STATS: AtomicBool = AtomicBool::new(false);
+
+pub fn enable_full_stats() {
+    ENABLE_FULL_STATS.store(true, Ordering::Release);
+}
+
+/// Keeps track of the number of times a task has been executed, and its
+/// duration.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum TaskStats {
+    Small(TaskStatsSmall),
+    Full(Box<TaskStatsFull>),
+}
+
+impl TaskStats {
+    /// Creates a new [`TaskStats`].
+    pub fn new() -> Self {
+        if ENABLE_FULL_STATS.load(Ordering::Acquire) {
+            Self::Full(Box::new(TaskStatsFull::default()))
+        } else {
+            Self::Small(TaskStatsSmall::default())
+        }
+    }
+
+    /// Resets the number of executions to 1 only if it was greater than 1.
+    pub fn reset_executions(&mut self) {
+        if let Self::Full(stats) = self {
+            if stats.executions > 1 {
+                stats.executions = 1;
+            }
+        }
+    }
+
+    /// Increments the number of executions by 1.
+    pub fn increment_executions(&mut self) {
+        if let Self::Full(stats) = self {
+            stats.executions += 1;
+        }
+    }
+
+    /// Registers a task duration.
+    pub fn register_duration(&mut self, duration: Duration) {
+        match self {
+            Self::Full(stats) => {
+                stats.total_duration += duration;
+                stats.last_duration = duration;
+            }
+            Self::Small(stats) => {
+                stats.last_duration = duration.as_nanos().try_into().unwrap_or(u64::MAX);
+            }
+        }
+    }
+
+    /// Resets stats to their default, zero-value.
+    pub fn reset(&mut self) {
+        match self {
+            Self::Full(stats) => {
+                stats.executions = 0;
+                stats.total_duration = Duration::ZERO;
+                stats.last_duration = Duration::ZERO;
+            }
+            Self::Small(stats) => {
+                stats.last_duration = 0;
+            }
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone, Eq, PartialEq)]
+pub struct TaskStatsSmall {
+    /// The last duration of the task in nanoseconds.
+    last_duration: u64,
+}
+
+impl TaskStatsSmall {
+    /// Returns the last duration of the task.
+    pub fn last_duration(&self) -> Duration {
```

Comment:
We don't actually use `last_duration` for non-viz purposes *yet*, but @sokra mentioned it will be used in GC. A u64 is smaller than a Duration (8 bytes vs. 16 bytes) and we don't need the additional precision.